Atomic append facilities

cehteh opened the pull request:
this is WIP I am still fleshing out some things. posting this early that you can see my intentions. Everything is still in flux.

You proposed 2 types one for non atomic and one for atomic access. Actually I need to do this in one type because i have some hybrid access. A non shared Vec can be mutated with &mut self, a shared Vec can be appended by a single thread from &self. When a &mut self is not available then .len() doing Relaxed ordering. This would be racy vs another thread that appends to the Vec but always memory safe when the append follows the contracts (only once thread append, length must be set after appending data). Clone does acquire semantic, since clone is already a expensive operation the cost is negligible and its the correct way when the atomic append is used.

A push_atomic/push_atomic_slice comes next.

Note: no rush: i said i want to use this in cowstr but that transition is a bit down in my queue.

cehteh committed:
WIP: make len atomic, add and use methods using atomics

This is a very early draft. method names and semantic may change.

cehteh committed:
Merge branch 'rust-cv:main' into main

cehteh committed:
rename methods in _exact/_strict, add 'atomic_append' feature, add 'spare_capacity()' method

cehteh committed:
add push_atomic()

cehteh commented:
so, I leave this now for review, probably for some weeks. This is still not completely finished, i will add a `push_slice()` and `push_slice_atomic()` method at least.

Whatever is finished is open for some bikeshedding about method names and whenever the atomic_append feature should be default or not. I don't have ARM hardware where it would make a measurable difference in performance. On x86 atomics are pretty free/nil and differences between with/without atomics are below the noise floor.
Because of this I decided to include the atomic_append in default. Even in embedded the majority of platforms nowadays support atomics.

On my side there is no urge to merge this yet. If this needs to be finished ping me.

vadixidav commented:
I see, so essentially `len` still has the same performance characteristics as the previous implementation? It looks like this still preserves the old behavior, just adding the new behavior for synchronization. I think we can stick to one type then. Thanks for the draft. I am looking forward to the final version being published and I can give a review then.

cehteh committed:
WIP: reserve/shrink API's

cehteh commented:
added reserve/reserve_exact/shrink_to/shrink_to_fit, reworked the resize_insert fn for that. The resize that can reservation of multiple elements is vital for adding slices efficiently (and i need it for other things). I see that i find time to add the slice things next when i find time.

cehteh committed:
FIX: the reseve functions need saturating add as well

cehteh committed:
FIX: Miri, unaligned access

* introduces a helper union to fix simplify alignment calculations
  no need for cmp and phantom data anymore
* add simple testcase that triggered the miri issue
* change end_ptr_atomic_mut(), using the rewritten start_ptr()

cehteh committed:
add: extend_from_slice()

cehteh committed:
add: extend_from_slice_atomic()

cehteh commented:
So, this concludes my work for now. Eventually other std::Vec mehtods/Traits (like Extend) could be added. Currently i have no urgent need for those. Integration into cowstr is kindof later in the queue i'll see if i miss some things then.

cehteh commented:
ping --- any news on my PR, do you intend to merge it?

vadixidav commented:
Sorry, I didn't realize it was out of draft. I will take a look at it later today. Thanks for the ping.

vadixidav commented:
I haven't forgotten, but yesterday was a bit busier than expected. I should be able to take care of this tonight.

cehteh commented:
On Wed, 18 Dec 2024 07:39:37 -0800
Geordon Worley ***@***.***> wrote:

> I haven't forgotten, but yesterday was a bit busier than expected. I
> should be able to take care of this tonight.

NP. i did the PR in hindight that I want to use it sometime next,
but there is no urge as i am currently busy with other things 
too. just wanted to ping you since i suspected it overseen it.
>

cehteh committed:
ADD: spare_capacity_mut() and set_len()

These methods have the same API/Semantic than the std::Vec methods.
They are useful when one wants to extend a HeaderVec in place in some complex way
like for example appending chars to a u8 headervec with `encode_utf8()`

cehteh commented on src/lib.rs:
```diff
@@ -297,11 +434,141 @@ impl<H, T> HeaderVec<H, T> {
+            // correct the len
+            let len_again = self.len_atomic_add_release(slice.len());
+            // in debug builds we check for races, the chance to catch these are still pretty minimal
+            #[cfg(debug_assertions)]
```
#[cfg(debug_assertions)] can be removed
may need #[allow(unused_variable)]

vadixidav commented:
Okay, today I will be getting to this. I will start reviewing now, but I will have to finish later today as I have something else.

vadixidav commented on src/lib.rs:
```diff
@@ -297,11 +436,150 @@ impl<H, T> HeaderVec<H, T> {
+    /// Before incrementing the length of the vector, you must ensure that new elements are
+    /// properly initialized.
+    #[inline(always)]
+    unsafe fn len_atomic_add_release(&self, n: usize) -> usize {
```
I am a bit confused by this method. I added the line about what it returns, but the `Safety` section is the same. From what it says, it is necessary for new elements to be initialized properly prior to this length being updated. Due to this, the only correct usage scenario where the original length could be different than the one you expected it to be is if two threads somehow coordinated the allocation of two new items in the vector, added their items, hit a barrier that blocks both of them from continuing until the items were added, then each subsequently atomically adds to the length in each thread. This is because if two threads are adding items, they must both synchronize the process of adding items to the vector, and therefore must already pre-agree upon what the length will ultimately be and what slot each thread will allocate into prior to incrementing the length to uphold the guarantee that looking inside the vector always results in valid initialized values. In this case there is no need for this function to return the size. In fact, it would make more sense for this to return nothing at all and instead be passed in the old size as an argument, panicking in the case that the returned value isn't equal to the expected old length.

My understanding might be incomplete though, but I am having a hard time understanding how this API would be used currently since the person adding the length must already be absolutely sure that all the items were added by all threads prior to it incrementing the length. Basically, two threads cant increment the length until some kind of synchronization has occurred to ensure the values were already added by the other threads into unique slots in the vector. The most likely use case would probably just be that there are N observers, but only 1 owner that can actually add items to the vector. In that case, I would suggest that we create an API to reflect that reality.

Let me know if I am misunderstanding how this is used.

cehteh committed:
add support for zero length HeaderVec

Having a HeaderVec being zero length is mostly useless because it has to reallocated instanly
when anything becomes pushed. This clearly should be avoided!

Nevertheless supporting zero length takes out a corner-case and a potential panic and removes
the burden for users explicitly ensuring zero length HeaderVecs don't happen in
practice. Generally improving software reliability.

cehteh committed:
DOC: fixing doc for reserve_cold()

cehteh committed:
Make HeaderVec being a NonNull<> instead a raw pointer

Since we always point to a valid allocation we can use NonNull here.
This will benefit from the niche optimization:
 size_of::<HeaderVec<H,T>>() == size_of::<Option<HeaderVec<H,T>>>

Also adds a test to verfiy that HeaderVec are always lean and niche optimized.

cehteh commented:
Note: I am not happy with the name `atomic_append` this was prolly a bit rushed. Since the  API it introduces only handles the length atomically. Pushes to immutable HeaderVecs need to properly synchronized still (thats why the atomic_push ops are unsafe). Any idea for a better name?

cehteh committed:
FIX: bug in offset() calculation

Off-By-One error when rouding to the next offset.
I totally missed that a bug slipped into the offset calculation because
I only tested with u8 and i32 data.

vadixidav commented:
Sorry, but I am actually a bit confused overall. I apologize for not fully comprehending the intention here.

My understanding is that you want to make the length atomic, and everything else remains the same. Given that concurrency issues can be a bit hard to comprehend, I would like to understand how a `std::vec::Vec` wrapped in a `std::sync::Mutex` is able to be mutated without any special atomic handling of the Vec's length and why that is different to this situation. Secondly, I would like to just confirm whether your intention is to actually have the push operation of the vector become atomic in some way, and if so what does that operationally look like. The reason is that I currently do not see how such an atomic push/append operation would be able to be safely executed using the given API. That isn't to say that it can't be done, but I am currently unable to understand. If we can explain how to safely use the unsafe APIs, we should document that process in the `# Safety` section in the docs. Specifically, we should explain how to prevent data races in where the data is being inserted into the vector, and how the atomic append operation can be executed simultaneously by two threads.

cehteh committed:
Document the atomic_append API / add Safety Section

cehteh committed:
remove is_empty_atomic_acquire as_slice_atomic_acquire

These public but undocumented and unused, i added them in hindsight but probably they are not
required. Otherwise they could be re-added later.

cehteh committed:
formatting

cehteh commented:
Note: meanwhile a lot stuff piled up in my PR's. Since I am using the things as i go everything is put on top of each other. Drop me a note when you want things factored out into single/smaller PR's or when I should change things before you want to move on (like using cfg_if)

cehteh committed:
improve conditional compilation w/o relying on cfg_if
